### ISSUE TITLE
TSC fails to locate .d.ts files when using "paths" remapping + ".js" extension

### ISSUE THREAD

**Issue by Nonanick** [microsoft/TypeScript#41642]
**TypeScript Version:**  4.0.5

**Search Terms:**
path remapping, paths, compilerOptions, .js, resolution, declaration types, .d.ts, module resolution, could not find declaration files

**Code**
Mocked File system:
```json
/ project / included / in /  file.ts  > File which shall require the ".js" file
/ project / root / of / file.js > Bundled file
/ project / root / of / file.d.ts > Bundled declarations file
```

in tsconfig.json:

```json
{
  "compilerOptions": {
  ...
  "moduleResolution": "node",
  "baseUrl": ".",
  "paths": {
      "/*": [
        "./*"
      ]
    }
  },
}
```
When running "tsc --traceResolution" the compiler successfully finds the .js file, resolving the module name, but does not lookup for .d.ts file:

```console
======== Resolving module '/root/of/file.js' from '/project/included/in/file.ts'. ========
Explicitly specified module resolution kind: 'NodeJs'.
'baseUrl' option is set to '/project', using this value to resolve non-relative module name '/root/of/file.js'.
'paths' option is specified, looking for a pattern to match module name '/root/of/file.js'.
Module name '/root/of/file.js', matched pattern '/*'.
Trying substitution './*', candidate module location: './root/of/file.js'.
File '/project/root/of/file.js' exist - use it as a name resolution result.
======== Module name '/root/of/file.js' was successfully resolved to '/project/root/of/file.js'. ========
```
As seen in the log messages TSC is able to resolve the .js file but does not load the "file.d.ts" standing right next to it!

**Expected behavior:**
When specifying a path ending with ".js", required for browser module resolution, Typescript should be able to locate the corresponding declaration file with a matching name on the same directory!
Removing the slash from the beginning, "root/of/file.js", makes Typescript correctly identify the .d.ts file, but browsers complain of non-relative paths...

**Actual behavior:**
Compiler / Typescript stops the resolution on the ".js" file failing to import type definitions therefore outputting the following error:
"Could not find a declaration file for module '/root/of/file.js'. '/project/root/of/file.js' implicitly has an 'any' type."

**Playground Link:** 
Cannot be replicated in the playground!

**Related Issues:** 
Could not find any open issue regarding the same subject!

**Comment by RyanCavanaugh** [comment]
This is intended since we can't tell if you're generating the declaration file from the JS file or not -- if you were, then preferring the .d.ts over the JS file would mean you never saw updates from the JS file

**Comment by Nonanick** [comment]
> This is intended since we can't tell if you're generating the declaration file from the JS file or not -- if you were, then preferring the .d.ts over the JS file would mean you never saw updates from the JS file

While I can agree with you when the "allowJS" option is set to true, this does not make sense when it's not;

Anyway what I've claimed is that the TS language server/compiler behaves/resolves differently when you use relative/absolute paths than when dealing with aliased paths. 
The alias resolution does not take into account the ".d.ts"/".ts" file while relative/absolute paths do!
This is, in my opinion, inconsistent! 

For the sake of simplicity I'll just paste both module resolutions using --traceResolution once again.

Using a relative path:
```
======== Resolving module '../lib/lib.js' from 'E:/resolution-issue/src/index.ts'. ========
Explicitly specified module resolution kind: 'NodeJs'.
Loading module as file / folder, candidate module location 'E:/resolution-issue/lib/lib.js', target file type 'TypeScript'.
File 'E:/resolution-issue/lib/lib.js.ts' does not exist.
File 'E:/resolution-issue/lib/lib.js.tsx' does not exist.
File 'E:/resolution-issue/lib/lib.js.d.ts' does not exist.
File name 'E:/resolution-issue/lib/lib.js' has a '.js' extension - stripping it.
File 'E:/resolution-issue/lib/lib.ts' 
does not exist.
File 'E:/resolution-issue/lib/lib.tsx' does not exist.
File 'E:/resolution-issue/lib/lib.d.ts' exist - use it as a name resolution 
result.
======== Module name '../lib/lib.js' was successfully resolved to 'E:/resolution-issue/lib/lib.d.ts'
```

Using an aliased path:
```
======== Resolving module '/lib/lib.js' from 'E:/resolution-issue/src/index.ts'. ========
Explicitly specified module resolution kind: 'NodeJs'.
'baseUrl' option is set to 'E:/resolution-issue', using this value to resolve non-relative module name '/lib/lib.js'.
'paths' option is specified, looking for a pattern to match module name '/lib/lib.js'.
Module name '/lib/lib.js', matched pattern '/*'.
Trying substitution './*', candidate module location: './lib/lib.js'.       
File 'E:/resolution-issue/lib/lib.js' 
exist - use it as a name resolution result.
======== Module name '/lib/lib.js' was successfully resolved to 'E:/resolution-issue/lib/lib.js'. ======== 
```

As you can see when using relative paths It tries to 'target file type' as Typescript and goes through a module resolution path, the same does not occur with aliased paths!

Even tough I'm using '/' as an alias the same resolution occurs if I prepend it with '@' or w/e!

* Proposal *
A way to flag an import so the compiler will use the "target file type 'TypeScript'" path instead of normal ".js" resolution ?

**Comment by RyanCavanaugh** [comment]
Can you post a complete example so we can be sure there's not some other configuration at play here?

**Comment by Nonanick** [comment]
Sure!
I'll be glad if If the whole issue is just a configuration tweak, love TS but some mechanisms are not that explicit in my head!

--------------------------------------

Folder structure:
![image](https://user-images.githubusercontent.com/3331237/101508438-70f4e380-3956-11eb-8595-1987d1d6b478.png)

* full 'tsconfig.json':
```json
  "compilerOptions": {
    "target": "es6",
    "module": "es6",
    "moduleResolution": "Node",
    "baseUrl": ".",
    "paths": {
      "/*": [
        "./*"
      ]
    },
    "declaration": true,
    "allowJs": false,
    "checkJs": false,
    "strict": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true
  }
}
```

* "/libs/lib.js":
```js
export class LibClass {
    foo() {
        return 'bar';
    }
}
```

* "/libs/lib.d.ts"
```ts
export declare class LibClass {
    foo(): string;
}
```

* "/src/lib.client.ts" using the aliased path
```ts
import { LibClass } from "/libs/lib.js";  // Could not find a declaration file for module '/libs/lib.js'. 'e:/resolution-issue/libs/lib.js' implicitly has an 'any' type.ts(7016)

let lib = new LibClass;
console.log(lib.foo());
```

* "/src/lib.client.ts" using the relative path
```ts
import { LibClass } from "../libs/lib.js"; // No issues, resolves the .d.ts file, Intelissense works, compiler won't complain

let lib = new LibClass;
console.log(lib.foo());
```

"tsc --version" outputs: 
- Version 4.1.2

I've already pasted "--traceResolution" for both cases.

## Notes  

1. Changing "allowJs" to true will fix the module resolution but it will break compilation with:
> "error TS5055: Cannot write file 'E:/resolution-issue/libs/lib.js' because it would overwrite input file."

Even when "/libs" is specified as an excluded directory;

tsconfig.json with "allowJs":
```json
{
  "compilerOptions": {
    "target": "es6",
    "module": "es6",
    "moduleResolution": "Node",
    "baseUrl": ".",
    "paths": {
      "/*": [
        "./*"
      ]
    },
    "declaration": true,
    "allowJs": true,
    "checkJs": false,
    "strict": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true
  },
  "include": [
    "src"
  ],
  "exclude": [
    "libs"
  ]
}
```

2. For now a workaround I found was to explicitly specify the resolution for my "lib.js" file as:
```json
{
  "compilerOptions": {
    "target": "es6",
    "module": "es6",
    "moduleResolution": "Node",
    "baseUrl": ".",
    "paths": {
      "/libs/lib.js": [
        "./libs/lib.d.ts",
        "./libs/lib.js"
      ]
    },
    "declaration": true,
    "allowJs": false,
    "checkJs": false,
    "strict": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true
  },
}
```
This fixes the module resolution and does not break compilation!

-----------------------------

If anything else is needed just ask, I will be glad to provide!

**Comment by sheetalkamat** [comment]
This is duplicate of #39743 and fixed by #42246

**Comment by imcm7** [comment]
I have the same with such paths

```
"paths": {
      "~/": [
        "./packages/"
      ],
      "~/*": [
        "./packages/*"
      ]
    }
```

in d.ts i get 

```
import { some } from '~/http/interfaces; // with error: Cannot find module '~/http/interfaces' or its corresponding type declarations.
```

Not understand what I do
